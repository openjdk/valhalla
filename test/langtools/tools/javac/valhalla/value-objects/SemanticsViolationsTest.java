/*
 * @test /nodynamiccopyright/
 * @bug 8279672
 * @summary Implement semantic checks for value classes
 * @compile/fail/ref=SemanticsViolationsTest.out -XDrawDiagnostics --should-stop=at=FLOW -XDdev SemanticsViolationsTest.java
 */

public class SemanticsViolationsTest {

    // A value class is implicitly final, so cannot be extended.
    value class Base {}
    class Subclass extends Base {} // Error: Base is implicitly final, cannot be extended.


    // The class may not be declared abstract.
    abstract value class AbsValue {}  // Error: value class cannot be abstract
    value interface ValueInterface {} // Error: interface cannot modified with value.

    // All instance fields are implicitly final, so must be assigned exactly
    // once by constructors or initializers, and cannot be assigned outside
    // of a constructor or initializer.
    value class Point {

        int x = 10;
        int y;
        int z;

        Point (int x, int y, int z) {
            this.x = x; // Error, final field 'x' is already assigned to.
            this.y = y; // OK.
            // Error, final z is unassigned.
        }
        void foo(Point p) {
            this.y = p.y; // Error, y is final and can't be written outside of ctor.
        }
    }

    // The class does not implement—directly or indirectly—IdentityObject.
    // This implies that the superclass is either Object or a stateless abstract class.
    value class IdentityValue implements IdentityObject { // Error, can't implement this
    }
    value class IdentityValue2 extends SemanticsViolationsTest { // Error, can't extend identity class
    }
    abstract static class AbstractWithState {
       int xx;
    }
    value class BrokenValue3 extends AbstractWithState { // Error, super class has state.
    }
    abstract static class AbstractWithoutState {
        static int ss;
    }
    value class GoodValue1 extends AbstractWithoutState {}
    value class GoodValue2 extends Object {} // allowed.

    // No constructor makes a super constructor call. Instance creation will
    // occur without executing any superclass initialization code.
    value class BrokenValue4 {
        BrokenValue4() {
            super(); // Error, can't chain to super's ctor.
        }
    }

    // No instance methods are declared synchronized.
    value class BrokenValue5 {
        synchronized void foo() {} // Error;
        synchronized static void soo() {} // OK.
        { synchronized(this) { /* Error.*/  } }
    }

    // The class does not declare a finalize() method.
    value class BrokenValue6 {
        public void finalize() {} // Error
    }

    // (Possibly) The constructor does not make use of this except to set
    // the fields in the constructor body, or perhaps after all fields are
    // definitely assigned.
    value class BrokenValue7 {
        int x;
        BrokenValue7() {
            foo(this); // Error.
            x = 10;
            foo(this); // Ok.
        }
        void foo(BrokenValue7 bv) {
        }
    }

    value record BrokenValue8(int x, int y) {
        synchronized void foo() { } // Error;
        synchronized static void soo() {} // OK.
    }
}
