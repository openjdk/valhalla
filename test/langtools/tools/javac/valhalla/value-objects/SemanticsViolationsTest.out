SemanticsViolationsTest.java:16:20: compiler.err.illegal.combination.of.modifiers: abstract, value
SemanticsViolationsTest.java:17:11: compiler.err.illegal.combination.of.modifiers: interface, value
SemanticsViolationsTest.java:12:28: compiler.err.cant.inherit.from.final: SemanticsViolationsTest.Base
SemanticsViolationsTest.java:65:27: compiler.err.mod.not.allowed.here: synchronized
SemanticsViolationsTest.java:90:27: compiler.err.mod.not.allowed.here: synchronized
SemanticsViolationsTest.java:12:5: compiler.err.identity.class.must.not.implement.value.object: SemanticsViolationsTest.Subclass
SemanticsViolationsTest.java:29:17: compiler.err.cant.assign.val.to.final.var: x
SemanticsViolationsTest.java:34:17: compiler.err.cant.assign.val.to.final.var: y
SemanticsViolationsTest.java:40:11: compiler.err.value.class.must.not.implement.identity.object: SemanticsViolationsTest.IdentityValue
SemanticsViolationsTest.java:42:11: compiler.err.concrete.supertype.for.value.class: SemanticsViolationsTest.IdentityValue2, SemanticsViolationsTest
SemanticsViolationsTest.java:47:11: compiler.err.super.field.not.allowed: xx, SemanticsViolationsTest.BrokenValue3, SemanticsViolationsTest.AbstractWithState
SemanticsViolationsTest.java:59:13: compiler.err.call.to.super.not.allowed.in.value.ctor
SemanticsViolationsTest.java:67:11: compiler.err.type.found.req: SemanticsViolationsTest.BrokenValue5, (compiler.misc.type.req.identity)
SemanticsViolationsTest.java:72:21: compiler.err.value.class.may.not.override: finalize
SemanticsViolationsTest.java:32:9: compiler.err.var.might.not.have.been.initialized: z
SemanticsViolationsTest.java:81:17: compiler.err.this.exposed.prematurely
SemanticsViolationsTest.java:81:16: compiler.err.this.exposed.prematurely
17 errors
